arch: arm: error: fix ARMv6-M assembly for Z_ARCH_EXCEPT

ioannisg opened the pull request:
As we are allowed to pass any integer value as as software
fatal exception reason, we need to fix the inline assembly
for ARMv6-M, to accept large immediate offsets. We do this
by changing MOV to LDR. We add a test case in kernel/fatal
to cover this scenario.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
ARMv6-M won't build for error reasons above 0xff.

agansari commented on include/arch/arm/error.h:
```diff
@@ -34,7 +34,7 @@ extern "C" {
 	__asm__ volatile ( \
 		"cpsie i\n\t" \
-		"movs r0, %[reason]\n\t" \
+		"ldr r0, =%[reason]\n\t" \
```
move to a separate commit

ioannisg replied:
Moved

andrewboie commented on tests/kernel/fatal/src/main.c:
```diff
@@ -297,6 +310,14 @@ void test_fatal(void)
+			NULL, NULL, NULL, K_PRIO_COOP(PRIORITY), 0,
+			K_NO_WAIT);
+	k_thread_abort(&alt_thread);
+	zassert_not_equal(rv, TC_FAIL, "thread was not aborted");
```
this test is run on all arches, so yes, otherwise CI would fail for this PR

ioannisg replied:
@andrewboie , I added a test case for arbitrary error reasons in SW-generated exceptions. There's a similar test for user mode (where only OOPS and ST FAIL may be reported)

ioannisg commented:
@andrewboie @agansari I will split the test-addition in a separate commit, and I will rework the Z_ARCH_EXCEPT for ARMv6, so it works in all cases. Stay tuned. This is the last prerequisite for USer Mode on ARMv6.m

ioannisg committed:
arch: arm: error: fix ARMv6-M assembly for Z_ARCH_EXCEPT

As we are allowed to pass any integer value as as software
fatal exception reason, we need to fix the inline assembly
for ARMv6-M, to accept large immediate offsets. We do this
by changing the way we write the exception reason to R0.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg committed:
tests: kernel: fatal: add a test-case for arbitrary error reason

We add a test-case in kernel/fatal test suite, to test that
the application developer can induce a SW-generated exception
with any 'reason' value.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>
